### PR TITLE
2.0.9 backported commits

### LINKED ISSUE #4404: [Opam beta 3] opam upgrades suggests beta variants by default
By default, `opam upgrade` wants to upgrade `ocaml` to its highest version (`4.13.0` as of now), even if this variant is in beta.
I think this is not a good idea in general. I don't know about it, but it may be due to the fact tham `opam` itself is in a beta version, but even then `ocaml-beta-repository` is not enabled, so I don't think it should be proposed.

### LINKED ISSUE #4549: opam list should not raise error 60 on packages flagged with `conf`
Using `OpamFileTools.lint ~check_upstream:true` or `opam lint` (opam 2.1.0~beta4), the following error was raised on a package flagged with `conf`:
```
$ cat opam
opam-version: "2.0"
maintainer: "https://github.com/ocaml/opam-repository/issues"
homepage: "https://asciidoc.org/"
bug-reports: "https://github.com/ocaml/opam-repository/issues"
authors: "Asciidoc dev team"
license: "GPL-2.0-only"
build: ["asciidoc" "--version"]
depexts: [
  ["asciidoc"] {os-family = "debian"}
  ["asciidoc"] {os-family = "ubuntu"}
  ["asciidoc"] {os-distribution = "fedora"}
  ["asciidoc"] {os-distribution = "centos"}
  ["asciidoc"] {os-distribution = "rhel"}
  ["asciidoc"] {os-distribution = "ol"}
  ["asciidoc"] {os-family = "suse"}
  ["asciidoc"] {os-family = "arch"}
  ["asciidoc"] {os-family = "alpine"}
  ["app-text/asciidoc"] {os-family = "gentoo"}
  ["asciidoc"] {os = "macos" & os-distribution = "homebrew"}
  ["asciidoc"] {os = "macos" & os-distribution = "macports"}
  ["asciidoc"] {os = "freebsd"}
  ["asciidoc"] {os = "openbsd"}
  ["asciidoc"] {os = "netbsd"}
]
synopsis: "Virtual package relying on asciidoc"
description:
  "This package can only install if the asciidoc binary is installed on the system."
flags: conf
```
```
$ opam lint --check-upstream
<path>/opam: Errors.
    error 60: Upstream check failed: "No url defined"
```

### LINKED ISSUE #4636: Introduce opam-root-version
Discussed on 1 Apr, introduce `opam-root-version` for the root and switch config files which is checked for taking out write locks. `opam-version` is checked for taking out read locks. So after root upgrade, `$OPAMROOT/config` would contain:

```
opam-version: "2.0"
opam-root-version: "2.1"
```

opam-state 2.0.x should be updated to internally recognise `opam-root-version` (i.e. no change in API in `OpamFile`) and ignore fields it doesn't recognise. The changes we've made for 2.1 mean that it should "just work" to have readable roots to 2.0.

As always, a given version of opam would never attempt to read a file with `opam-version` greater than itself.

### PR DESCRIPTION
**Pull request by rjbou**
Backported commits for the 2.0.9 release

Note: [blocking](https://github.com/ocaml/opam/pull/4535#issuecomment-778353834) setup-ocaml w/dune

## Still to cherry-pick
- [x] https://github.com/ocaml/opam/pull/4668
- [x] https://github.com/ocaml/opam/pull/4707

## Error free opam root loading
  * Adapted mechanism for 2.0, no upgrade [#4638 @rjbou - fix #4636]

## Sandbox
  * Fix the conflict with the environment variable name used by dune [#4535 @smorimoto - fix ocaml/dune#4166]
  * Kill builds on Ctrl-C with bubblewrap [#4530 @kit-ty-kate - fix #4404]
  * Linux: mount existing TMPDIR read-only, re-bind `$TMPDIR` to a separate tmpfs [#4589 @AltGr]
  * Fix the sandbox check [#4589 @AltGr]
  * Fix sandbox script shell mistake that made `PWD` read-write on remove actions [#4589 @AltGr]
  * Port bwrap improvements to sandbox_exec [#4589 #4609 @AltGr]
 
## Lint
  * Fix W59 & E60 with conf flag handling (no url required) [#4550 @rjbou - fix #4549]

## Internal
  * Fix temporary file with a too long name causing errors on Windows [#4590 @AltGr]
  * Add missing shell quoting to support space and special shell characters in switch directory path [#4707 @kit-ty-kate]

## Package building
  * The options `--root` and `--switch` are now reflected in environment variables when building packages so that calls to `opam` during build access the correct root and switch [#4668 @LasseBlaauwbroek]

## Build
  * Switch to newer version of MCCS (based on newer GLPK) for src_ext [#4559 @AltGr]
  * Fix opam-devel's tests on platforms without openssl, GNU-diff and a system-wide ocaml [#4500 @kit-ty-kate]
 * Untag dune as a build dependency [#4229 @kit-ty-kate]
 * Build release tags in the same as release archives [#4666 @dra27]

## Pin
  * Fix version pin source retrieving: don't error if archive opam file is malformed [#4580 @rjbou]
  
## Infrastructure
  * Release scripts: switch to OCaml 4.10.2 by default, add macos/arm64 builds by default [#4559 @AltGr]

## Test
  * Fix configure check in github actions [#4593 @rjbou]

### CODE REVIEW

**Comment by smorimoto** [comment]
Need to replace some `@` with `#`.

```diff
- - Fix sandbox script shell mistake that made PWD read-write on remove actions [@4589 @AltGr]
- - Port bwrap improvements to sandbox_exec [@4589 @AltGr]
+ - Fix sandbox script shell mistake that made PWD read-write on remove actions [#4589 @AltGr]
+ - Port bwrap improvements to sandbox_exec [#4589 @AltGr]
```

File: CHANGES [hunk 1]
```diff
@@ -2,6 +2,28 @@ Changes prefixed with "(*)" are potentially breaking to scripts or existing
...
+2.0.9:
+* Fix the conflict with the environment variable name used by dune
+  [#4535 @smorimoto - fix ocaml/dune#4166]
+* Kill builds on Ctrl-C with bubblewrap [#4530 @kit-ty-kate - fix #4404]
+* Linux: mount existing TMPDIR read-only, re-bind `$TMPDIR` to a separate tmpfs
+  [#4589 @AltGr]
+* Fix the sandbox check [#4589 @AltGr]
+* Fix sandbox script shell mistake that made `PWD` read-write on remove actions
+  [@4589 @AltGr]
+* Port bwrap improvements to sandbox_exec [@4589 @AltGr]
```
Comment (rjbou) [inline CHANGES:14]:
```suggestion
  [#4589 @AltGr]
* Port bwrap improvements to sandbox_exec [#4589 @AltGr]
```

**Comment by rjbou** [comment]
indeed, thanks!

**Review by kit-ty-kate — APPROVED** [review]

**Comment by dra27** [comment]
Linking in https://github.com/ocaml/opam/issues/4573

**Comment by smorimoto** [comment]
Can't we release 2.0.9 without #4573? There are some important changes for macOS and I hope to release 2.0.9 in May.

**Comment by dra27** [comment]
We're trying to get the 2.1 release candidate out of the door, first - once that's done (which really should be any day now!), releasing 2.0.9 would be fine. It's not blocked on 4573, it was just linked to consider it.

The solutions to #4394 and #4636 both need to be back-ported to 2.0.x, although not necessarily to 2.0.9. See [the milestone](https://github.com/ocaml/opam/milestone/42) for more info.

**Comment by smorimoto** [comment]
That sounds good. Thank you for the perfect explanation!

**Comment by rjbou** [comment]
Last update, rebased and commits squashed!

**Comment by smorimoto** [comment]
Yay!

**Comment by dra27** [comment]
`src_ext/dune-opam-file-format` needs to be nuked

File: src/format/opamFile.ml [hunk 2]
```diff
@@ -1029,8 +1079,13 @@ module ConfigSyntax = struct
 
   let internal = "config"
 
+  let format_version = OpamVersion.of_string "2.0"
+  let file_format_version = OpamVersion.of_string "2.0"
```
Comment (dra27) [inline src/format/opamFile.ml:1083]: These are unused in the back-port

File: src/format/opamFile.ml [hunk 3]
```diff
@@ -1202,16 +1264,39 @@ module ConfigSyntax = struct
       (fun (fld, ppacc) -> fld, Pp.embed with_wrappers wrappers ppacc)
       Wrappers.fields
 
-  let pp =
+  let pp_cond ?f ?condition () =
     let name = internal in
     Pp.I.map_file @@
+    Pp.I.check_opam_version () -|
```
Comment (dra27) [inline src/format/opamFile.ml:1270]: The 2.1 version pass `~f` here - the optional argument above wants removing for the 2.0 version (it's triggering a warning)

File: src/format/opamFile.ml [hunk 4]
```diff
@@ -1457,16 +1542,19 @@ module Repos_configSyntax = struct
          OpamRepositoryName.Map.(of_list, bindings));
   ]
 
-  let pp =
+  let pp_cond ?f ?condition () =
```
Comment (dra27) [inline src/format/opamFile.ml:1545]:
```suggestion
  let pp_cond ?condition () =
```

File: src/format/opamFile.ml [hunk 5]
```diff
@@ -1535,11 +1623,13 @@ module Switch_configSyntax = struct
          fld, Pp.embed (fun wrappers t -> {t with wrappers}) (fun t -> t.wrappers) ppacc)
       Wrappers.fields
 
-  let pp =
+  let pp_cond ?f ?condition () =
```
Comment (dra27) [inline src/format/opamFile.ml:1626]:
```suggestion
  let pp_cond ?condition () =
```

File: src/format/opamFile.ml [hunk 6]
```diff
@@ -1604,18 +1708,21 @@ module SwitchSelectionsSyntax = struct
        Pp.of_pair "Package set" OpamPackage.Set.(of_list, elements))
   ]
 
-  let pp =
+  let pp_cond ?f ?condition () =
```
Comment (dra27) [inline src/format/opamFile.ml:1711]:
```suggestion
  let pp_cond ?condition () =
```

**Review by dra27 — COMMENTED** [review]
A few warnings from OCaml which are worth eliminating (even on an LTS branch!)

**Pull request merged**